monitoring/v1/types.go

```diff
@@ -0,0 +1,337 @@
+// than the upstream type in order to allow better OpenShift specific defaulting
+// and validation, while not modifying the upstream APIs directly.
+//
+// You can find upstream API documentation for PrometheusRule resources here:
+//
+// https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+type AlertingRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the desired state of this AlertingRule object.
+	Spec AlertingRuleSpec `json:"spec"`
+
+	// status describes the current state of this AlertOverrides object.
+	//
+	// +optional
+	Status AlertingRuleStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlertingRuleList is a list of AlertingRule objects.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +k8s:openapi-gen=true
+type AlertingRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertingRule objects.
+	Items []AlertingRule `json:"items"`
+}
+
+// AlertingRuleSpec is the desired state of an AlertingRule resource.
+//
+// +k8s:openapi-gen=true
+type AlertingRuleSpec struct {
+	// groups is a list of grouped alerting rules.  Rule groups are the unit at
+	// which Prometheus parallelizes rule processing.  All rules in a single group
+	// share a configured evaluation interval.  All rules in the group will be
+	// processed together on this interval, sequentially, and all rules will be
+	// processed.
+	//
+	// It's common to group related alerting rules into a single AlertingRule
+	// resources, and within that resource, closely related alerts, or simply
+	// alerts with the same interval, into individual groups.  You are also free
+	// to create AlertingRule resources with only a single rule group, but be
+	// aware that this can have a performance impact on Prometheus if the group is
+	// extremely large or has very complex query expressions to evaluate.
+	// Spreading very complex rules across multiple groups to allow them to be
+	// processed in parallel is also a common use-case.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems:=1
+	Groups []RuleGroup `json:"groups"`
+}
+
+// RuleGroup is a list of sequentially evaluated alerting rules.
+//
+// +k8s:openapi-gen=true
+type RuleGroup struct {
+	// name is the name of the group.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```

Comment:
Did we conclude there was any way to restrict this by character set?

Reply:
Where is this being passed to? Does it end up in logs? Does it end up in any prometheus configuration? Is there anywhere where this could be injected that might mean that a user could do something unintended (thinking injection attacks here) if we allow any character?

Have you tested that required is the same as `minLength: 1`, because I suspect that required will pass if `name: ""` (you could use our integration testing suite to verify this

Reply:
> Where is this being passed to? Does it end up in logs? Does it end up in any prometheus configuration? Is there anywhere where this could be injected that might mean that a user could do something unintended (thinking injection attacks here) if we allow any character?

This ends up in the Prometheus rules configuration file. I can't think of any way this could trigger a vulnerability. If one exists then it is on the Prometheus side.

> Have you tested that required is the same as minLength: 1, because I suspect that required will pass if name: "" (you could use our integration testing suite to verify this)

Pretty sure that required ≠ minLengh=1.

Reply:
+1 on having minLength and required. The group name must not be empty. [Prometheus validates this](https://github.com/prometheus/prometheus/blob/8f3b2fc9aeaf1bd3f839e820e7f9ffc571006e89/model/rulefmt/rulefmt.go#L101-L109) as well. I don't think we can validate the "unique rule group names per file" requirement here.

I agree with Simon on the vulnerability question. The group name is and should only ever be treated as a string. Prometheus-operators [PrometheusRule CRD also uses a string](https://github.com/prometheus-operator/prometheus-operator/blob/6c18db1083546ae6950f5fda33bd6629c10487e2/pkg/apis/monitoring/v1/prometheusrule_types.go#L63) here.

Reply:
> I don't think we can validate the "unique rule group names per file" requirement here

it will be handled by `+listType=map` and `+listMapKey=name` on the `groups` field.